core/src/mast/serialization/mod.rs

```diff
@@ -232,35 +229,105 @@ impl Deserializable for MastForest {
+            let node_id = MastNodeId::try_from((node_id, &mast_forest))?;
+            mast_forest.set_decorators(node_id, decorator_list);
+        }
+
         // read "before enter" and "after exit" decorators, and update the corresponding nodes
         let before_enter_decorators: Vec<(usize, Vec<DecoratorId>)> =
             read_before_after_decorators(source, &mast_forest)?;
         for (node_id, decorator_ids) in before_enter_decorators {
-            let node_id: u32 = node_id.try_into().map_err(|_| {
-                DeserializationError::InvalidValue(format!(
-                    "Invalid node id '{node_id}' while deserializing"
-                ))
-            })?;
-            let node_id = MastNodeId::from_u32_safe(node_id, &mast_forest)?;
+            let node_id = MastNodeId::try_from((node_id, &mast_forest))?;
             mast_forest.set_before_enter(node_id, decorator_ids);
         }
 
         let after_exit_decorators: Vec<(usize, Vec<DecoratorId>)> =
             read_before_after_decorators(source, &mast_forest)?;
         for (node_id, decorator_ids) in after_exit_decorators {
-            let node_id: u32 = node_id.try_into().map_err(|_| {
-                DeserializationError::InvalidValue(format!(
-                    "Invalid node id '{node_id}' while deserializing"
-                ))
-            })?;
-            let node_id = MastNodeId::from_u32_safe(node_id, &mast_forest)?;
+            let node_id = MastNodeId::try_from((node_id, &mast_forest))?;
             mast_forest.set_after_exit(node_id, decorator_ids);
         }
 
         Ok(mast_forest)
     }
 }
 
+fn read_and_validate_magic<R: ByteReader>(source: &mut R) -> Result<[u8; 5], DeserializationError> {
+    let magic: [u8; 5] = source.read_array()?;
+    if magic != *MAGIC {
+        return Err(DeserializationError::InvalidValue(format!(
+            "Invalid magic bytes. Expected '{:?}', got '{:?}'",
+            *MAGIC, magic
+        )));
+    }
+    Ok(magic)
+}
+
+fn read_and_validate_version<R: ByteReader>(
+    source: &mut R,
+) -> Result<[u8; 3], DeserializationError> {
+    let version: [u8; 3] = source.read_array()?;
+    if version != VERSION {
+        return Err(DeserializationError::InvalidValue(format!(
+            "Unsupported version. Got '{version:?}', but only '{VERSION:?}' is supported",
+        )));
+    }
+    Ok(version)
+}
+
+fn read_block_decorators<R: ByteReader>(
+    source: &mut R,
+) -> Result<Vec<(usize, DecoratorList)>, DeserializationError> {
+    let vec_len: usize = source.read()?;
+    let mut out_vec: Vec<_> = Vec::with_capacity(vec_len);
+
+    for _ in 0..vec_len {
+        let node_id: usize = source.read()?;
+
+        let decorator_vec_len: usize = source.read()?;
+        let mut inner_vec: Vec<(usize, DecoratorId)> = Vec::with_capacity(decorator_vec_len);
+        for _ in 0..decorator_vec_len {
+            let op_id: usize = source.read()?;
+            let decorator_id: DecoratorId = source.read()?;
```

Comment:
I think we should do this analogous to `read_before_after_decorators` and use `let decorator_id = DecoratorId::from_u32_safe(source.read()?, mast_forest)?;` so we have the additional safety check that the `DecoratorId` we're reading points to a valid decorator and is not out of bounds.
That way we don't need `impl Deserializable for DecoratorId` as well.